newrelic/resource_newrelic_data_partition.go

```diff
@@ -0,0 +1,281 @@
+					value := v.(string)
+					var diags diag.Diagnostics
+					if !strings.HasPrefix(value, "Log_") {
+						diag := diag.Diagnostic{
+							Severity: diag.Error,
+							Summary:  "INVALID_DATA_PARTITION_INPUT",
+							Detail:   "Prepend \"Log_\" to the given target_data_partition value.",
+						}
+						diags = append(diags, diag)
+					}
+					return diags
+				},
+			},
+			"deleted": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Whether or not this data partition rule is deleted. Deleting a data partition rule does not delete the already persisted data. This data will be retained for a given period of time specified in the retention policy field.",
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Second),
+		},
+	}
+}
+
+func listValidDataPartitionRuleRetentionPolicyType() []string {
+	return []string{
+		string(logconfigurations.LogConfigurationsDataPartitionRuleRetentionPolicyTypeTypes.SECONDARY),
+		string(logconfigurations.LogConfigurationsDataPartitionRuleRetentionPolicyTypeTypes.STANDARD),
+	}
+}
+
+// Create the data partition rule
+func resourceNewRelicDataPartitionCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	providerConfig := meta.(*ProviderConfig)
+	client := providerConfig.NewClient
+
+	accountID := selectAccountID(providerConfig, d)
+
+	createInput := logconfigurations.LogConfigurationsCreateDataPartitionRuleInput{
+		Description: d.Get("description").(string),
+		Enabled:     d.Get("enabled").(bool),
+		NRQL:        logconfigurations.NRQL(d.Get("nrql").(string)),
+	}
+
+	//The name of a log data partition. Has to start with 'Log_' prefix and can only contain alphanumeric characters and underscores.
+	if e, ok := d.GetOk("target_data_partition"); ok {
+		createInput.TargetDataPartition = logconfigurations.LogConfigurationsLogDataPartitionName(e.(string))
+	}
+
+	if e, ok := d.GetOk("retention_policy"); ok {
+		createInput.RetentionPolicy = logconfigurations.LogConfigurationsDataPartitionRuleRetentionPolicyType(e.(string))
+	}
+	log.Printf("[INFO] Creating New Relic Data Partition Rule  %s", createInput.TargetDataPartition)
+
+	created, err := client.Logconfigurations.LogConfigurationsCreateDataPartitionRuleWithContext(ctx, accountID, createInput)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	var apiDiags diag.Diagnostics
+
+	//Setting the errors
+	if created.Errors != nil {
+		for _, err := range created.Errors {
+			apiDiags = append(apiDiags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  err.Message,
+				Detail:   string(err.Type),
+			})
+		}
+		return apiDiags
+	}
+
+	if created == nil {
+		return diag.Errorf("err: data partition rule create result wasn't returned or rule was not created.")
+	}
+
+	ruleID := created.Rule.ID
+
+	d.SetId(ruleID)
+
+	//Need retry mechanism
+	retryErr := resource.RetryContext(ctx, d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
```

Comment:
Just curious, do we need the retry due to async latency behavior?

Reply:
Yes